### PR TITLE
Converted the DRPC mechanism to use binary (ByteBuffer) results sets but preserve String output for existing users.

### PR DESCRIPTION
**Pull request by jasonkolb**
executeBinary returns java.nio.ByteBuffer, legacy "execute" method now
converts from ByteBuffer to String using default encoding

### CODE REVIEW

**Comment by tomjack** [comment]
sweet!

File: src/clj/backtype/storm/daemon/drpc.clj [hunk 1]
```diff
@@ -70,8 +71,36 @@
               (throw result)
               result
               ))))
+      (^String execute [this ^String function ^String args]
```
Comment (nathanmarz) [inline src/clj/backtype/storm/daemon/drpc.clj:74]: Need to get rid of the code duplication here. Execute should be a simple wrapper around executeBinary

File: src/jvm/backtype/storm/ILocalDRPC.java [hunk 2]
```diff
@@ -3,8 +3,16 @@
...
 import backtype.storm.generated.DistributedRPC;
 import backtype.storm.generated.DistributedRPCInvocations;
+import backtype.storm.generated.DRPCExecutionException;
 
+import org.apache.thrift7.TException;
+import java.nio.charset.CharacterCodingException;
 
 public interface ILocalDRPC extends DistributedRPC.Iface, DistributedRPCInvocations.Iface, Shutdownable {
     public String getServiceId();    
+
```
Comment (nathanmarz) [inline src/jvm/backtype/storm/ILocalDRPC.java:13]: What is this?

File: src/jvm/backtype/storm/drpc/DRPCInvocationsClient.java [hunk 3]
```diff
@@ -48,6 +52,29 @@ public void result(String id, String result) throws TException {
         }
     }
 
+    /**
+    * Overloaded method which accepts a string as a result and converts it to a ByteBuffer in the background
+    **/
+    public void result(String id, String result) throws TException, java.nio.charset.CharacterCodingException {
+        try {
```
Comment (nathanmarz) [inline src/jvm/backtype/storm/drpc/DRPCInvocationsClient.java:59]: Need to get rid of code duplication here. This should be a wrapper around result(String, ByteBuffer)

File: src/jvm/backtype/storm/drpc/ReturnResults.java [hunk 4]
```diff
@@ -33,7 +36,28 @@ public void prepare(Map stormConf, TopologyContext context, OutputCollector coll
 
     @Override
     public void execute(Tuple input) {
-        String result = (String) input.getValue(0);
+
+        ByteBuffer result = null;
+
+        if( input.getValue(0) instanceof ByteBuffer )
```
Comment (nathanmarz) [inline src/jvm/backtype/storm/drpc/ReturnResults.java:42]: I believe this is dependent on #199 being implemented.

Also, I think you should also handle the case of getting a regular byte array too.

File: src/clj/backtype/storm/daemon/drpc.clj [hunk 5]
```diff
@@ -70,8 +71,36 @@
...
+          (let [result (@id->result id)]
+            (cleanup id)
+            (log-debug "Returning DRPC result for " function " " args " at " (System/currentTimeMillis))
+            (if (instance? DRPCExecutionException result)
+              (throw result)
+
+              ;(-> (.newDecoder (Charset/defaultCharset)) (.decode result) str)
+              ;str
+
+              (-> (Charset/defaultCharset)
```
Comment (nathanmarz) [inline src/clj/backtype/storm/daemon/drpc.clj:96]: The String encoding/decoding stuff should be wrapped in a utility function (probably in Utils.java)

**Comment by nathanmarz** [comment]
@jasonkolb I left some comments. I also think adding a short test to test the binary code paths in drpc_test.clj would be good.

**Comment by nathanmarz** [comment]
Closing this since there's been no activity on this pull request and there's still work that needs to be done. I'd be happy to accept another pull request that completes this feature.

**Pull request closed without merging**